oshmem/mca/scoll/ucc/scoll_ucc_module.c

```diff
@@ -0,0 +1,357 @@
+
+#define OBJ_RELEASE_IF_NOT_NULL( obj ) if( NULL != (obj) ) OBJ_RELEASE( obj );
+
+int mca_scoll_ucc_init_query(bool enable_progress_threads, bool enable_ucc_threads)
+{
+    return OSHMEM_SUCCESS;
+}
+
+static void mca_scoll_ucc_module_clear(mca_scoll_ucc_module_t *ucc_module)
+{
+    ucc_module->previous_barrier      = NULL;
+    ucc_module->previous_broadcast    = NULL;
+    ucc_module->previous_reduce       = NULL;
+    ucc_module->previous_collect      = NULL;
+    ucc_module->previous_alltoall     = NULL;
+}
+
+static void mca_scoll_ucc_module_construct(mca_scoll_ucc_module_t *ucc_module)
+{
+    mca_scoll_ucc_module_clear(ucc_module);
+}
+
+int mca_scoll_ucc_progress(void)
+{
+    ucc_context_progress(mca_scoll_ucc_component.ucc_context);
+    return OSHMEM_SUCCESS;
+}
+
+static void mca_scoll_ucc_module_destruct(mca_scoll_ucc_module_t *ucc_module)
+{
+    OBJ_RELEASE_IF_NOT_NULL(ucc_module->previous_alltoall_module);
+    OBJ_RELEASE_IF_NOT_NULL(ucc_module->previous_collect_module);
+    OBJ_RELEASE_IF_NOT_NULL(ucc_module->previous_reduce_module);
+    OBJ_RELEASE_IF_NOT_NULL(ucc_module->previous_broadcast_module);
+    OBJ_RELEASE_IF_NOT_NULL(ucc_module->previous_barrier_module);
+
+    mca_scoll_ucc_module_clear(ucc_module);
+}
+
+#define UCC_SAVE_PREV_SCOLL_API(__api) do {\
+    ucc_module->previous_ ## __api            = osh_group->g_scoll.scoll_ ## __api;\
+    ucc_module->previous_ ## __api ## _module = osh_group->g_scoll.scoll_ ## __api ## _module;\
+    if (!osh_group->g_scoll.scoll_ ## __api || !osh_group->g_scoll.scoll_ ## __api ## _module) {\
+        UCC_VERBOSE(1, "no underlying " # __api"; disqualifying myself");\
+        return OSHMEM_ERROR;\
+    }\
+    OBJ_RETAIN(ucc_module->previous_ ## __api ## _module);\
+} while(0)
+
+static int mca_scoll_ucc_save_coll_handlers(mca_scoll_base_module_t *module, oshmem_group_t *osh_group)
+{
+    mca_scoll_ucc_module_t* ucc_module = (mca_scoll_ucc_module_t*) module;
+    UCC_SAVE_PREV_SCOLL_API(barrier);
+    UCC_SAVE_PREV_SCOLL_API(broadcast);
+    UCC_SAVE_PREV_SCOLL_API(reduce);
+    UCC_SAVE_PREV_SCOLL_API(collect);
+    UCC_SAVE_PREV_SCOLL_API(alltoall);
+    return OSHMEM_SUCCESS;
+}
+
+typedef struct oob_allgather_req 
+{
+    void           *sbuf;
+    void           *rbuf;
+    void           *oob_coll_ctx;
+    size_t          msglen;
+} oob_allgather_req_t;
+
+static ucc_status_t oob_allgather_free(void *req)
+{
+    free(req);
+    return UCC_OK;
+}
+
+static ucc_status_t oob_allgather(void *sbuf, void *rbuf, size_t msglen,
+                                  void *oob_coll_ctx, void **req)
+{
+    oob_allgather_req_t *oob_req = malloc(sizeof(*oob_req));
+    oob_req->sbuf                = sbuf;
+    oob_req->rbuf                = rbuf;
+    oob_req->msglen              = msglen;
+    oob_req->oob_coll_ctx        = oob_coll_ctx;
+    *req                         = oob_req;
+    return UCC_OK;
+}
+
+static ucc_status_t oob_allgather_test(void *req)
```

Comment:
This oob allgather looks like a blocking implementation - ucc.h spec requires non-blocking oob allgather if i recall. @manjugv plz comment. This might work currently due to implementation, but in general, inside libucc we treat oob.allgather as non-blocking fn.

Reply:
Yes, oob.allgather in UCC is a non-blocking function. @wfaderhold21 You might to adapt to it.